File: src/test/java/com/heekng/celloct/repository/ManagerRepositoryTest.java
package com.heekng.celloct.repository;

import com.heekng.celloct.entity.Manager;
import com.heekng.celloct.entity.Member;
import com.heekng.celloct.entity.Shop;
import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.EntityManager;

import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.*;
import static org.junit.jupiter.api.Assertions.*;

@SpringBootTest
@Transactional
class ManagerRepositoryTest {

    @Autowired
    ManagerRepository managerRepository;
    @Autowired
    MemberRepository memberRepository;
    @Autowired
    ShopRepository shopRepository;
    @Autowired
    EntityManager em;

    @Test
    void create() throws Exception {
        //given
        Member member = Member.builder()
                .name("member1")
                .email("member1@heekng.com")
                .build();
        memberRepository.save(member);

        Shop shop = Shop.builder()
                .phone("010-1234-1234")
                .name("shop1")
                .build();
        shopRepository.save(shop);

        Manager manager = Manager.builder()
                .member(member)
                .shop(shop)
                .name("managerName")
                .build();
        managerRepository.save(manager);

        em.flush();
        em.clear();

        //when
        Manager findManager = managerRepository.findById(manager.getId()).get();
        //then
        assertThat(findManager.getMember().getId()).isEqualTo(manager.getMember().getId());
        assertThat(findManager.getShop().getId()).isEqualTo(manager.getShop().getId());
    }

    @Test
    void delete() throws Exception {
        //given
        Member member = Member.builder()
                .name("member1")
                .email("member1@heekng.com")
                .build();
        memberRepository.save(member);

        Shop shop = Shop.builder()
                .phone("010-1234-1234")
                .name("shop1")
                .build();
        shopRepository.save(shop);

        Manager manager = Manager.builder()
                .member(member)
                .shop(shop)
                .name("managerName")
                .build();
        managerRepository.save(manager);
        em.flush();
        em.clear();
        //when
        Manager findManager = managerRepository.findById(manager.getId()).get();
        managerRepository.delete(findManager);
        em.flush();
        em.clear();

        Optional<Manager> managerOptional = managerRepository.findById(manager.getId());
        //then
        assertThat(managerOptional).isEmpty();
    }

    @Test
    void findByMemberIdAndShopIdTest() throws Exception {
        //given
        Member member = Member.builder()
                .name("member1")
                .email("member1@heekng.com")
                .build();
        memberRepository.save(member);

        Shop shop = Shop.builder()
                .phone("010-1234-1234")
                .name("shop1")
                .build();
        shopRepository.save(shop);

        Manager manager = Manager.builder()
                .member(member)
                .shop(shop)
                .name("managerName")
                .build();
        managerRepository.save(manager);
        em.flush();
        em.clear();
        //when
        Optional<Manager> managers = managerRepository.findByMemberIdAndShopId(member.getId(), shop.getId());

        //then
        assertThat(managers.get().getName()).isEqualTo(manager.getName());
    }

    @Test
    void findByMemberIdTest() throws Exception {
        //given
        Member member = Member.builder()
                .name("member1")
                .email("member1@heekng.com")
                .build();
        memberRepository.save(member);

        Shop shop = Shop.builder()
                .phone("010-1234-1234")
                .name("shop1")
                .build();
        shopRepository.save(shop);

        Manager manager = Manager.builder()
                .member(member)
                .shop(shop)
                .name("managerName")
                .build();
        managerRepository.save(manager);
        em.flush();
        em.clear();
        //when
        List<Manager> managers = managerRepository.findByMemberId(member.getId());

        //then
        assertThat(managers.get(0).getName()).isEqualTo(manager.getName());
    }

    @Test
    void findWithShopByMemberIdTest() throws Exception {
        //given
        Member member = Member.builder()
                .name("member1")
                .email("member1@heekng.com")
                .build();
        memberRepository.save(member);

        Shop shop = Shop.builder()
                .phone("010-1234-1234")
                .name("shop1")
                .build();
        shopRepository.save(shop);

        Manager manager = Manager.builder()
                .member(member)
                .shop(shop)
                .name("managerName")
                .build();
        managerRepository.save(manager);
        em.flush();
        em.clear();
        //when
        List<Manager> managers = managerRepository.findWithShopByMemberId(member.getId());
        em.flush();
        em.clear();

        //then
        assertThat(managers.get(0).getShop().getName()).isEqualTo(shop.getName());
    }

    @Test
    void findListByShopIdTest() throws Exception {
        //given
        Member member = Member.builder()
                .name("member1")
                .email("member1@heekng.com")
                .build();
        memberRepository.save(member);

        Shop shop = Shop.builder()
                .phone("010-1234-1234")
                .name("shop1")
                .build();
        shopRepository.save(shop);

        Manager manager = Manager.builder()
                .member(member)
                .shop(shop)
                .name("managerName")
                .build();
        managerRepository.save(manager);
        em.flush();
        em.clear();
        //when
        List<Manager> managers = managerRepository.findListByShopId(shop.getId());
        //then
        assertThat(managers.get(0).getName()).isEqualTo(manager.getName());
    }

    @Test
    void findWithMemberByIdTest() throws Exception {
        //given
        Member member = Member.builder()
                .name("member1")
                .email("member1@heekng.com")
                .build();
        memberRepository.save(member);

        Shop shop = Shop.builder()
                .phone("010-1234-1234")
                .name("shop1")
                .build();
        shopRepository.save(shop);

        Manager manager = Manager.builder()
                .member(member)
                .shop(shop)
                .name("managerName")
                .build();
        managerRepository.save(manager);
        em.flush();
        em.clear();
        //when
        Manager findManager = managerRepository.findWithMemberById(manager.getId());
        em.flush();
        em.clear();
        //then
        assertThat(findManager.getMember().getName()).isEqualTo(member.getName());
    }

    @Test
    void findByShopIdAndIdTest() throws Exception {
        //given
        Member member = Member.builder()
                .name("member1")
                .email("member1@heekng.com")
                .build();
        memberRepository.save(member);

        Shop shop = Shop.builder()
                .phone("010-1234-1234")
                .name("shop1")
                .build();
        shopRepository.save(shop);

        Manager manager = Manager.builder()
                .member(member)
                .shop(shop)
                .name("managerName")
                .build();
        managerRepository.save(manager);
        em.flush();
        em.clear();
        //when
        Optional<Manager> managers = managerRepository.findByShopIdAndId(shop.getId(), manager.getId());

        //then
        assertThat(managers.get().getName()).isEqualTo(manager.getName());
    }

}

File: src/test/kotlin/com/heekng/celloct/repository/ManagerRepositoryTest.kt
package com.heekng.celloct.repository

import com.heekng.celloct.entity.Manager
import com.heekng.celloct.entity.Member
import com.heekng.celloct.entity.Shop
import com.heekng.celloct.util.findByIdOrThrow
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.data.repository.findByIdOrNull
import org.springframework.transaction.annotation.Transactional
import javax.persistence.EntityManager

@SpringBootTest
@Transactional
class ManagerRepositoryTest @Autowired constructor(
    private val managerRepository: ManagerRepository,
    private val memberRepository: MemberRepository,
    private val shopRepository: ShopRepository,
    private val em: EntityManager,
) {

    @Test
    fun createTest() {
        //given
        val member = Member.fixture("member1", "member1@heekng.com")
        memberRepository.save(member)
        val shop = Shop.fixture("010-1234-1234", "shop1", null)
        shopRepository.save(shop)
        val manager = Manager.fixture(shop, member, "managerName")
        managerRepository.save(manager)
        em.flush()
        em.clear()
        //when
        val findManager = managerRepository.findByIdOrThrow(manager.id)
        //then
        assertThat(findManager.id).isEqualTo(manager.id)
        assertThat(findManager.member.id).isEqualTo(member.id)
        assertThat(findManager.shop.id).isEqualTo(shop.id)
    }

    @Test
    fun deleteTest() {
        //given
        val member = Member.fixture("member1", "member1@heekng.com")
        memberRepository.save(member)
        val shop = Shop.fixture("010-1234-1234", "shop1", null)
        shopRepository.save(shop)
        val manager = Manager.fixture(shop, member, "managerName")
        managerRepository.save(manager)
        em.flush()
        em.clear()
        //when
        val findManager = managerRepository.findByIdOrThrow(manager.id)
        managerRepository.delete(findManager)
        em.flush()
        em.clear()
        val findManagerOrNull = managerRepository.findByIdOrNull(manager.id)
        //then
        assertThat(findManagerOrNull).isNull()
    }

    @Test
    fun findByMemberIdAndShopIdTest() {
        //given
        val member = Member.fixture("member", "member@heekng.com")
        memberRepository.save(member)
        val shop = Shop.fixture("010-1234-1234", "shop", null)
        shopRepository.save(shop)
        val manager = Manager.fixture(shop, member, "managerName")
        managerRepository.save(manager)
        em.flush()
        em.clear()
        //when
//        val findManager = managerRepository.findByMemberIdAndShopId(member.id!!, shop.id!!)
        val findManager = managerRepository.find(
            memberId = member.id,
            shopId = shop.id
        )[0]
        //then
        assertThat(findManager.id).isEqualTo(manager.id)
    }

    @Test
    fun findByMemberIdTest() {
        //given
        val member = Member.fixture("member", "member@heekng.com")
        memberRepository.save(member)
        val shop = Shop.fixture("010-1234-1234", "shop", null)
        shopRepository.save(shop)
        val manager = Manager.fixture(shop, member, "managerName")
        managerRepository.save(manager)
        em.flush()
        em.clear()
        //when
        val findManagers = managerRepository.find(memberId = member.id)
        //then
        assertThat(findManagers).hasSize(1)
        assertThat(findManagers[0].id).isEqualTo(manager.id)
    }

    @Test
    fun findWithShopByMemberIdTest() {
        //given
        val member = Member.fixture("member", "member@heekng.com")
        memberRepository.save(member)
        val shop = Shop.fixture("010-1234-1234", "shop", null)
        shopRepository.save(shop)
        val manager = Manager.fixture(shop, member, "managerName")
        managerRepository.save(manager)
        em.flush()
        em.clear()
        //when
        val managers = managerRepository.findWithShopByMemberId(member.id!!)
        em.flush()
        em.clear()
        //then
        assertThat(managers).hasSize(1)
        assertThat(managers[0].id).isEqualTo(manager.id)
    }

    @Test
    fun findListByShopIdTest() {
        //given
        val member = Member.fixture("member", "member@heekng.com")
        memberRepository.save(member)
        val shop = Shop.fixture("010-1234-1234", "shop", null)
        shopRepository.save(shop)
        val manager = Manager.fixture(shop, member, "managerName")
        managerRepository.save(manager)
        em.flush()
        em.clear()
        //when
        val managers = managerRepository.find(shopId = shop.id)
        //then
        assertThat(managers).hasSize(1)
        assertThat(managers[0].id).isEqualTo(manager.id)
    }

    @Test
    fun findWithMemberByIdTest() {
        //given
        val member = Member.fixture("member", "member@heekng.com")
        memberRepository.save(member)
        val shop = Shop.fixture("010-1234-1234", "shop", null)
        shopRepository.save(shop)
        val manager = Manager.fixture(shop, member, "managerName")
        managerRepository.save(manager)
        em.flush()
        em.clear()
        //when
        val findManager = managerRepository.findWithMemberById(manager.id!!)
        em.flush()
        em.clear()
        //then
        assertThat(findManager.id).isEqualTo(manager.id)
    }

    @Test
    fun findByShopIdAndIdTest() {
        //given
        val member = Member.fixture("member", "member@heekng.com")
        memberRepository.save(member)
        val shop = Shop.fixture("010-1234-1234", "shop", null)
        shopRepository.save(shop)
        val manager = Manager.fixture(shop, member, "managerName")
        managerRepository.save(manager)
        em.flush()
        em.clear()
        //when
        val findManager = managerRepository.findByShopIdAndId(shop.id!!, manager.id!!)
        //then
        assertThat(findManager!!.id).isEqualTo(manager.id)
    }
}